chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala

```diff
@@ -25,14 +25,22 @@ sealed abstract class Aggregate extends Data {
     }
 
     // Check that children obey the directionality rules.
-    val childDirections = getElements.map(_.direction).toSet
+    val childDirections = this match {
+      case vec: Vec[_] =>
+        // Use direction of sample_element, must bind it to get an ActualDirection
+        vec.sample_element.bind(SampleElementBinding(vec), resolvedDirection)
+        Set(vec.sample_element.direction)
+      case rec: Record =>
+        getElements.collect({
+          case elt: Element => elt.direction
+          case vec: Vec[_] => vec.direction
+          // Don't check the directions of empty Records (see https://github.com/freechipsproject/chisel3/pull/946)
+          case rec: Record if rec.getElements.nonEmpty => rec.direction
```

Comment:
I can't quite say what the larger implications of this will be, but the inspecting the contents of this's elements seems like an odd code smell. Perhaps it might make sense to have an ActualDirection,NoDirection, because that's kind of what we're getting at here fundamentally?